src/bootstrap/src/core/build_steps/test.rs

```diff
@@ -680,7 +680,16 @@ impl Step for Miri {
             .arg("--manifest-path")
             .arg(builder.src.join("src/tools/miri/test-cargo-miri/Cargo.toml"));
         cargo.arg("--target").arg(target.rustc_target_arg());
-        cargo.arg("--tests"); // don't run doctests, they are too confused by the staging
+
+        if stage == 0 {
+            builder.info("INFO: miri/test-cargo-miri doctests are ignored on stage 0.");
+            // Don't run doctests on stage0, stage0 rustdoc is incompatible with the miri sysroot
+            cargo.arg("--tests");
+        } else {
+            // Miri doesn't know which rustdoc to use by default, so tell which one it should use.
+            cargo.env("RUSTDOC_REAL", builder.rustdoc(compiler));
```

Comment:
Have you figured out why `prepare_tool_cargo` does not set this?

Reply:
But why is want_rustdoc not set? It should be set whenever we are running `./x.py test` without `--no-doc`...

Reply:
The actual problem is that `cmd` is `run`. Since we are using `cargo run` to run `cargo-miri miri test`, it doesn't "see" that this is a test invocation.

Reply:
Oh, right. Now it's even more complicated to handle.

Reply:
It's fine, we want a rustdoc from a different stage anyway so we couldn't use the value it sets.